### ISSUE TITLE
[Project]: Add Driver Drowsiness Detection from Facial Landmarks using MediaPipe + LSTM (Real-Time Webcam)

### ISSUE THREAD

**Issue by divyanshim27** [abhisheks008/DL-Simplified#1199]
## Summary

I would like to propose and implement: **Driver Drowsiness Detection using Facial Landmarks and LSTM.** While DL-Simplified already has a `Distracted Driver Detection` and `Drowsiness Detection using EEG signals` project, there is no project using real-time webcam-based facial landmark analysis for drowsiness — the most practical and deployable approach for real-world driver assistance. This project uses MediaPipe FaceMesh to extract Eye Aspect Ratio (EAR) and Mouth Aspect Ratio (MAR) as temporal sequences, which are then classified by an LSTM to detect microsleep and yawning events.

## Why This Project

- This is a genuinely deployable safety application — not just a classifier on a static dataset.
- It introduces contributors to temporal sequence modeling (LSTM) applied to a regression output from computer vision (MediaPipe).
- The combination of MediaPipe + LSTM for safety-critical real-time inference is a technique used in production ADAS systems.
- It demonstrates real-time inference in a Gradio web app, which is more engaging for learners than a static notebook.

## Proposed Implementation

**Core drowsiness metric extraction:**
```python
import mediapipe as mp
import numpy as np

def eye_aspect_ratio(landmarks, eye_indices):
    """
    Compute EAR for one eye using 6 MediaPipe FaceMesh landmark indices.
    EAR < 0.25 for >3 frames indicates eye closure (drowsiness).
    """
    p = [(landmarks[i].x, landmarks[i].y) for i in eye_indices]
    # Vertical distances
    A = np.linalg.norm(np.array(p[1]) - np.array(p[5]))
    B = np.linalg.norm(np.array(p[2]) - np.array(p[4]))
    # Horizontal distance
    C = np.linalg.norm(np.array(p[0]) - np.array(p[3]))
    return (A + B) / (2.0 * C)

# LEFT_EYE = [33, 160, 158, 133, 153, 144]
# RIGHT_EYE = [362, 385, 387, 263, 373, 380]
```

**LSTM Classifier on EAR/MAR temporal sequences:**
```python
import torch
import torch.nn as nn

class DrowsinessLSTM(nn.Module):
    def __init__(self, input_size=2, hidden_size=64, num_layers=2, num_classes=3):
        super().__init__()
        self.lstm = nn.LSTM(input_size, hidden_size, num_layers,
                            batch_first=True, dropout=0.3)
        self.fc = nn.Linear(hidden_size, num_classes)  # alert / drowsy / yawning

    def forward(self, x):
        out, _ = self.lstm(x)
        return self.fc(out[:, -1, :])  # classify based on last timestep
```

**Dataset:** Trained on sliding window sequences (30 frames × 2 features: EAR, MAR) extracted from the UTA-RLDD dataset (publicly available, 60 subjects).

**Notebook structure:**

Driver Drowsiness Detection using DL/
├── Dataset/
│ └── README.md # UTA-RLDD download link and preprocessing
├── Images/
│ ├── ear_mar_plot.png
│ ├── mediapipe_landmarks.png
│ ├── lstm_architecture.png
│ └── confusion_matrix.png
├── Model/
│ ├── Driver_Drowsiness_Detection.ipynb
│ └── README.md
├── Web App/
│ ├── app.py # Gradio real-time webcam app
│ └── README.md
└── requirements.txt

## Additional Notes

- Real-time inference runs at ~25fps on CPU — no GPU required for deployment.
- The Gradio app will access the user's webcam and show live EAR/MAR readings alongside the drowsiness classification.
- Three models compared: (1) Threshold-only EAR/MAR rules, (2) Random Forest on features, (3) LSTM on sequences.

Could you please assign this issue to me?

**Labels:** `enhancement`, `new-project`, `computer-vision`, `lstm`, `real-time`, `help wanted`, `GSSoC 2026`
